lib/db/observed.go

```diff
@@ -0,0 +1,265 @@
+	delete(dl, device)
+}
+
+func (dl DropListObserved) MarkFolder(folder string, devices []protocol.DeviceID) {
+	for _, dev := range devices {
+		folders := dl.MarkDevice(dev)
+		if folders == nil {
+			folders = make(map[string]struct{})
+			dl[dev] = folders
+		}
+		folders[folder] = struct{}{}
+	}
+}
+
+// shouldDropPendingDevice defines how the drop-list is interpreted for pending devices
+func (dl DropListObserved) shouldDropPendingDevice(key []byte, keyer keyer) bool {
+	keyDev := keyer.DeviceFromPendingDeviceKey(key)
+	//FIXME: DeviceIDFromBytes() panics when given a wrong length input.
+	//       It should rather return an error which we'd check for here.
+	if len(keyDev) != protocol.DeviceIDLength {
+		l.Warnf("Invalid pending device entry, deleting from database: %x", key)
+		return true
+	}
+	// Valid entries are looked up in the drop-list, invalid ones cleaned up
+	deviceID := protocol.DeviceIDFromBytes(keyDev)
+	_, dropDevice := dl[deviceID]
+	if dropDevice {
+		l.Debugf("Removing marked pending device %v", deviceID)
+	}
+	return dropDevice
+}
+
+// shouldDropPendingFolder defines how the drop-list is interpreted for pending folders,
+// which is different and more nested than for devices
+func (dl DropListObserved) shouldDropPendingFolder(key []byte, keyer keyer) bool {
+	keyDev, ok := keyer.DeviceFromPendingFolderKey(key)
+	if !ok {
+		l.Warnf("Invalid pending folder entry, deleting from database: %x", key)
+		return true
+	}
+	// Valid entries are looked up in the drop-list, invalid ones cleaned up
+	deviceID := protocol.DeviceIDFromBytes(keyDev)
+	dropFolders, allowDevice := dl[deviceID]
+	// Check the associated set of folders if provided, otherwise drop.
+	if !allowDevice {
+		l.Debugf("Removing pending folder offered by %v", deviceID)
+		return true
+	}
+	if len(dropFolders) == 0 {
+		// Map is empty or nil, skip lookup
+		return false
+	}
+	folderID := keyer.FolderFromPendingFolderKey(key)
+	// Drop only mentioned folder IDs
+	_, dropFolder := dropFolders[string(folderID)]
+	if dropFolder {
+		l.Debugf("Removing marked pending folder %s for %v", folderID, deviceID)
+	}
+	return dropFolder
+}
+
+// CleanPendingDevices removes all pending device entries matching a given set of device IDs
+func (db *Lowlevel) CleanPendingDevices(dropList DropListObserved) {
+	iter, err := db.NewPrefixIterator([]byte{KeyTypePendingDevice})
+	if err != nil {
+		l.Warnf("Could not iterate through pending device entries for cleanup: %v", err)
+		return
+	}
+	defer iter.Release()
+	for iter.Next() {
+		if dropList.shouldDropPendingDevice(iter.Key(), db.keyer) {
+			if err := db.Delete(iter.Key()); err != nil {
+				l.Warnf("Failed to remove pending device entry: %v", err)
+			}
+		}
+	}
+}
+
+// CleanPendingFolders removes all pending folder entries not matching a given set of
+// device IDs, or matching the set of folder IDs associated with those given devices.
+func (db *Lowlevel) CleanPendingFolders(dropList DropListObserved) {
```

Comment:
Similarly I think `CleanPendingFolders(device protocol.DeviceID, folders []string)` and one will have to call it for each device that is known to have folders...

Reply:
It also needs to be called for devices that are no longer mentioned anywhere else (got removed or ignored) so it's impossible to make a positive list of pending device / folder combinations to drop.